### PR TITLE
feat: enable NRF response caching

### PR DESCRIPTION
**Pull request by dariofaccin**
This PR aims to implement NRF response caching in AUSF module.

### CODE REVIEW

**Comment by gab-arrobo** [comment]
@thakurajayL, please help me review this PR. Thanks!

File: producer/functions.go [hunk 1]
```diff
@@ -277,7 +277,7 @@ func getUdmUrl(nrfUri string) string {
 	nfDiscoverParam := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{
 		ServiceNames: optional.NewInterface([]models.ServiceName{models.ServiceName_NUDM_UEAU}),
 	}
-	res, err := consumer.SendSearchNFInstances(nrfUri, models.NfType_UDM, models.NfType_AUSF, nfDiscoverParam)
+	res, err := consumer.SendSearchNFInstances(nrfUri, models.NfType_UDM, models.NfType_AUSF, &nfDiscoverParam)
```
[the remark this answers is not in the document]
Comment (gatici) [reply producer/functions.go:280]: In other way, setting a shorter CacheEvictionInterval may solve the issue for above situations.

File: consumer/nf_management.go [hunk 2]
```diff
@@ -137,3 +137,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
...
+	client := Nnrf_NFManagement.NewAPIClient(configuration)
+	var res *http.Response
+
+	res, err = client.SubscriptionIDDocumentApi.RemoveSubscription(context.Background(), subscriptionId)
+	if err == nil {
+		return
+	} else if res != nil {
+		defer func() {
+			if bodyCloseErr := res.Body.Close(); bodyCloseErr != nil {
+				err = fmt.Errorf("RemoveSubscription' response body cannot close: %+w", bodyCloseErr)
```
Comment (gatici) [inline consumer/nf_management.go:187]:
```suggestion
				err = fmt.Errorf("RemoveSubscription' response body cannot close: %w", bodyCloseErr)
```
To log the errors  %w is used.

File: consumer/nf_discovery.go [hunk 3]
```diff
@@ -10,29 +10,62 @@ import (
...
 	configuration.SetBasePath(nrfUri)
 	client := Nnrf_NFDiscovery.NewAPIClient(configuration)
 
-	result, rsp, rspErr := client.NFInstancesStoreApi.SearchNFInstances(context.TODO(),
-		targetNfType, requestNfType, &param)
-	if rspErr != nil {
-		return nil, fmt.Errorf("NFInstancesStoreApi Response error: %+w", rspErr)
+	result, res, err := client.NFInstancesStoreApi.SearchNFInstances(context.TODO(), targetNfType, requestNfType, param)
+	if res != nil && res.StatusCode == http.StatusTemporaryRedirect {
+		err = fmt.Errorf("Temporary Redirect For Non NRF Consumer")
```
Comment (gatici) [inline consumer/nf_discovery.go:40]:
```suggestion
		err = fmt.Errorf("temporary redirect for non NRF consumer")
```
Error logs should not be capitalized.
https://go.dev/wiki/CodeReviewComments#error-strings

**Review by gatici — COMMENTED** [review]
We need to add unit test for the new added functions.

**Comment by gatici** [comment]
Generally looks good. Thank you.

**Review by gab-arrobo — COMMENTED** [review]
Hi @dariofaccin, have you tested the changes? I am asking this because I just finished to run a test using `Aether-OnRamp` and as you can see below, all profiles failed due to "registration reject" (`procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT"`)

```
time="2024-08-08T17:26:43Z" level=info msg="Profile Name: profile1 , Profile Type: register"
time="2024-08-08T17:26:43Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:43Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T17:26:43Z" level=info msg="Profile Name: profile2 , Profile Type: pdusessest"
time="2024-08-08T17:26:43Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:43Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T17:26:44Z" level=info msg="Profile Name: profile3 , Profile Type: anrelease"
time="2024-08-08T17:26:44Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:44Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T17:26:44Z" level=info msg="Profile Name: profile4 , Profile Type: uetriggservicereq"
time="2024-08-08T17:26:44Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:44Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T17:26:45Z" level=info msg="Profile Name: profile5 , Profile Type: deregister"
time="2024-08-08T17:26:45Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:45Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T17:26:45Z" level=info msg="Profile Name: profile7 , Profile Type: uereqpdusessrelease"
time="2024-08-08T17:26:45Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T17:26:45Z" level=info msg="Profile Status: FAIL"
```

File: consumer/nf_discovery.go [hunk 4]
```diff
@@ -10,29 +10,69 @@ import (
 	"fmt"
 	"net/http"
 
+	ausf_context "github.com/omec-project/ausf/context"
```
Comment (gab-arrobo) [inline consumer/nf_discovery.go:13]: Follow the same naming convention already used in this file
```suggestion
	ausfContext "github.com/omec-project/ausf/context"
```

File: consumer/nf_discovery.go [hunk 5]
```diff
@@ -10,29 +10,69 @@ import (
...
+	CreateSubscription        = SendCreateSubscription
+	NRFCacheSearchNFInstances = nrfCache.SearchNFInstances
+	StoreApiClient            = &Nnrf_NFDiscovery.APIClient{}
+	StoreApiSearchNFInstances = (*Nnrf_NFDiscovery.NFInstancesStoreApiService).SearchNFInstances
 )
 
-func SendSearchNFInstances(nrfUri string, targetNfType, requestNfType models.NfType,
-	param Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (*models.SearchResult, error) {
+var SendSearchNFInstances = func(nrfUri string, targetNfType, requestNfType models.NfType,
+	param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts,
```
Comment (gab-arrobo) [inline consumer/nf_discovery.go:28]: Does this "comma" have to be here?

File: consumer/nf_management.go [hunk 6]
```diff
@@ -137,3 +137,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
...
+	} else {
+		err = openapi.ReportError("server no response")
+	}
+	return
+}
+
+var SendRemoveSubscription = func(subscriptionId string) (problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Infoln("Send Remove Subscription")
+
+	ausfSelf := ausf_context.GetSelf()
```
Comment (gab-arrobo) [inline consumer/nf_management.go:174]: Is it possible to rename this import (`ausf_context` to `ausfContext`) throughout the file?

File: consumer/nf_management.go [hunk 7]
```diff
@@ -137,3 +137,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
...
+		problem := err.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
+		problemDetails = &problem
+	} else {
+		err = openapi.ReportError("server no response")
+	}
+	return
+}
+
+var SendRemoveSubscription = func(subscriptionId string) (problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Infoln("Send Remove Subscription")
```
Comment (gab-arrobo) [inline consumer/nf_management.go:172]:
```suggestion
	logger.ConsumerLog.Infoln("send Remove Subscription")
```

File: consumer/nf_management.go [hunk 8]
```diff
@@ -137,3 +137,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
 	}
 	return
 }
+
+var SendCreateSubscription = func(nrfUri string, nrfSubscriptionData models.NrfSubscriptionData) (nrfSubData models.NrfSubscriptionData, problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Debugf("Send Create Subscription")
```
Comment (gab-arrobo) [inline consumer/nf_management.go:142]:
```suggestion
	logger.ConsumerLog.Debugf("send Create Subscription")
```

File: consumer/nf_management.go [hunk 9]
```diff
@@ -137,3 +137,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
...
+	client := Nnrf_NFManagement.NewAPIClient(configuration)
+	var res *http.Response
+
+	res, err = client.SubscriptionIDDocumentApi.RemoveSubscription(context.Background(), subscriptionId)
+	if err == nil {
+		return
+	} else if res != nil {
+		defer func() {
+			if bodyCloseErr := res.Body.Close(); bodyCloseErr != nil {
+				err = fmt.Errorf("RemoveSubscription' response body cannot close: %w", bodyCloseErr)
```
Comment (gab-arrobo) [inline consumer/nf_management.go:187]:
```suggestion
				err = fmt.Errorf("RemoveSubscription's response body cannot close: %w", bodyCloseErr)
```

File: context/ausf_context_init.go [hunk 10]
```diff
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"os"
 	"strconv"
+	"time"
 
 	"github.com/google/uuid"
```
Comment (gab-arrobo) [inline context/ausf_context_init.go:16]:
```suggestion
```

File: producer/callback.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/omec-project/ausf/logger"
+	"github.com/omec-project/openapi/models"
+	nrfCache "github.com/omec-project/openapi/nrfcache"
+	"github.com/omec-project/util/httpwrapper"
+)
+
+var NRFCacheRemoveNfProfileFromNrfCache = nrfCache.RemoveNfProfileFromNrfCache
+
+func HandleNfSubscriptionStatusNotify(request *httpwrapper.Request) *httpwrapper.Response {
+	logger.ProducerLog.Traceln("Handle NF Status Notify")
```
Comment (gab-arrobo) [inline producer/callback.go:23]:
```suggestion
	logger.ProducerLog.Traceln("handle NF Status Notify")
```

File: producer/callback.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	if notificationData.Event == models.NotificationEventType_DEREGISTERED {
+		if ausfContext.GetSelf().EnableNrfCaching {
+			ok := NRFCacheRemoveNfProfileFromNrfCache(nfInstanceId)
+			logger.ProducerLog.Tracef("nfinstance %v deleted from cache: %v", nfInstanceId, ok)
+		}
+		if subscriptionId, ok := ausfContext.GetSelf().NfStatusSubscriptions.Load(nfInstanceId); ok {
+			logger.ConsumerLog.Debugf("SubscriptionId of nfInstance %v is %v", nfInstanceId, subscriptionId.(string))
+			problemDetails, err := consumer.SendRemoveSubscription(subscriptionId.(string))
+			if problemDetails != nil {
+				logger.ConsumerLog.Errorf("Remove NF Subscription Failed Problem[%+v]", problemDetails)
```
Comment (gab-arrobo) [inline producer/callback.go:60]:
```suggestion
				logger.ConsumerLog.Errorf("remove NF Subscription Failed Problem[%+v]", problemDetails)
```

File: producer/callback.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+			} else if err != nil {
+				logger.ConsumerLog.Errorf("Remove NF Subscription Error[%+v]", err)
```
Comment (gab-arrobo) [inline producer/callback.go:62]:
```suggestion
				logger.ConsumerLog.Errorf("remove NF Subscription Error[%+v]", err)
```

File: producer/callback.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+			} else {
+				logger.ConsumerLog.Infoln("Remove NF Subscription successful")
```
Comment (gab-arrobo) [inline producer/callback.go:64]:
```suggestion
				logger.ConsumerLog.Infoln("remove NF Subscription successful")
```

**Comment by dariofaccin** [comment]
@gab-arrobo I updated the PR with the fixes for the crash and the suggestions. Now it is ready for review again

**Review by gab-arrobo — APPROVED** [review]
+1

**Pull request merged**